### ISSUE TITLE
how save latent code?

### ISSUE THREAD

**Issue by molo32** [giannisdaras/ilo#2]
how save latent code after optimization?

**Comment by giannisdaras** [comment]
Hi,
try setting save_latent=True to the config (https://github.com/giannisdaras/ilo/blob/master/configs/config.yaml#L104), or equivalently, run the script with the argument `stylegan.save_latent=1`.

**Comment by molo32** [comment]
I already got it but how do I open it again?

`np.load ("/ content / ilo / files / latent.pt"). files` returns
`['archive / data.pkl', 'archive / data / 218201040', 'archive / version']`

Which of all is the latent code and how open it ?  I want to make an interpolation between two latents

**Comment by giannisdaras** [comment]
This how I usually restore saved latents: https://github.com/giannisdaras/ilo/blob/master/ilo_stylegan.py#L124
Let me know if it worked!

**Comment by molo32** [comment]
Thanks for answering, but now I have another problem, when I load the 3 files that they saved after inverting, and I try to reverse and save the image, it is not the same as the previously inverted image.

```
%cd /content/ilo
from model import Generator
gen = Generator(1024, 512, 8)


import torchvision
import numpy as np
import math
import torch
from torch import optim
from torch.nn import functional as F
from torchvision import transforms
from PIL import Image
from tqdm import tqdm
import torch.nn as nn
import lpips
import matplotlib.pyplot as plt

#config=dataMap
gen.load_state_dict(torch.load("/content/ilo/stylegan2-ffhq-config-f.pt")["g_ema"], strict=False)
gen.eval()
gen = gen.to("cuda")
gen.start_layer = 0
gen.end_layer = 8


class MappingProxy(nn.Module):
    def __init__(self,gaussian_ft):
        super(MappingProxy,self).__init__()
        self.mean = gaussian_ft["mean"]
        self.std = gaussian_ft["std"]
        self.lrelu = torch.nn.LeakyReLU(0.2)
    def forward(self,x):
        x = self.lrelu(self.std * x + self.mean)
        return x
mpl = MappingProxy(torch.load('/content/ilo/gaussian_fit.pt'))

latent_z= mpl(torch.load('/content/ilo/files/latent.pt').to('cuda'))
noisee=torch.load('/content/ilo/files/noises.pt')
layer_inn=torch.load('/content/ilo/files/gen_outs.pt')

img_gen, _ =  gen([latent_z] ,
                                  input_is_latent=True,
                  noise=noisee,
                       layer_in=layer_inn             )


torchvision.utils.save_image(
                    img_gen,
                    f'fake.png',
                    nrow=int(img_gen.shape[0] ** 0.5),
                    normalize=True)


```

with this code I load the 3 files latent_z noise and latent_in, then I reverse the image again but it is not like the inverted image above.
What am I doing wrong?
this image better explains what i mean



![problem](https://user-images.githubusercontent.com/55426197/109409784-3b1f3400-7974-11eb-87de-627e5a91a187.png)

**Comment by giannisdaras** [comment]
Before the forward pass to the generator, specify the start_layer. I.e, try:
`gen.start_layer=3` (if you have optimized up to 3rd layer for example)
and then: `img_gen, _ = gen([latent_z], ...) `

Let me know if it worked :)

**Comment by molo32** [comment]
It works but I would like to know if I can convert the gen.start_layer = 3 layer into a single latent code such as (1,18,512).
   without using  (gen.start_layer = "x", layer_inn , and noise), in the original nvidia stylegan2 generator.
as in the original generator, you just pass the latent code and synthesize the image

**Comment by giannisdaras** [comment]
You cannot convert it to a single latent code.
The idea in ILO is that there are three things that control the output: the latents, the noises AND the output of the previous layers. So, you are optimizing over three different things and that's why you need to save three vectors.
You can easily do interpolation though, just mix the latents, as you would do for StyleGAN.

**Comment by giannisdaras** [comment]
Closing due to inactivity, feel free to reopen if you have additional questions.

**Comment by JohnnyRisk** [comment]
@giannisdaras Thank you so much for your hard work on this project. I am very impressed with the results and the reconstructions that ilo is able to achieve. Just a follow-up question to be sure I understand what is going on here. Is it in fact the case that we are effectively overriding the weights of the generator up until the layer that we optimize to? So the reason that @molo32 is getting different results is because we use gen_output to override any output that would have come previously? I am looking at your code for and layer_in seems to do exactly that, and that would be the reason that the start_layer has to start at the layer in which you ended your optimization? If this is the case, would the result be the same for @molo32's case if you were to run the optimization, save the generator with all its weights, save the noises, and the latents then run it? If not could you please explain why that would not work. I am thinking about the use case where you first want to reconstruct an image and then you want to use the latents to edit it in some projected direction. Thank you again, for your hard work.

**Comment by giannisdaras** [comment]
Hi @JohnnyRisk!
Thanks for your nice comments! 
you understand this correctly, basically gen_out overrides any output that would have come previously.

> If this is the case, would the result be the same for @molo32's case if you were to run the optimization, save the generator with all its weights, save the noises, and the latents then run it?

I am not sure I am parsing this correctly, but for sure you can get back the result of the optimization by saving the latents and the gen out and then recovering them and feeding them to the model with start_layer= the layer you ended the ILO optimization. Now, if you want to not use gen_out, then you will end up with a different image than the ILO output. Let me know if that makes sense!

**Comment by JohnnyRisk** [comment]
Thank you for the quick response. I reread my post and I am not sure the second part of my question was clear. Lol. Let me try again. My question is about the reason that you need the gen_out. Is this because we have changed the generator parameters of the first n layers. So, to make this more concrete, lets say in order to generate a new image we optimize the first 3 layers. Lets call the original generator parameters w and the optimized parameters w*. So, is the issue that we need to w1*, w2* and w3* which are different from w1, w2, w3. And because of this when we just load w and the latents and the noises for all layers (w1-w8) we will get a different gen_output for layer 3. So the way that you get around this is you over ride the output from w1-w3 with gen_output3 which would actually correspond to w1*,w2*, and w3*? If this is the case what I am wondering is if instead of saving gen_output3 and using that, if I were to save the new generator weights w1*,w2*,w3*,w4,w5,w6,w7,w8 and use the latents and noises would i get the same thing as the generated output from running ilo?

**Comment by JohnnyRisk** [comment]
The reason I am asking this question is because, one way that this work may be used in practice is to first reconstruct and image and then by investigating the generator, edit the image in a way that was desirable. However, it appears that some things that one may want to be able to edit are in the first n layers that we optimize and then effectively freeze. If this is not the case perhaps we can have a discussion offline, I would love to understand this more deeply.

**Comment by JohnnyRisk** [comment]
again @giannisdaras thank you so much for the great work! I find it very interesting and a very nice contribution to the literature.

**Comment by giannisdaras** [comment]
@JohnnyRisk we do not optimize the parameters of the generator, we optimize the inputs of the generator (parameters are frozen). So, imagine you cut the generator into two pieces, G1 and G2. The output of G2 depends on the output of G1 and some additional inputs (noises and latents for G2). Now, because G1 cannot generate all outputs, we throw away G1 and we feed to G2 gen_out, which is a variable we optimize over starting with the output of G1 (hence the name). Happy to discuss this more offline if you have more questions :)

**Comment by JohnnyRisk** [comment]
Ahhh thank you, that makes perfect sense
